Bug fixes for netcdf-c cmake build

climbfuji opened the pull request:
## Description

This PR fixes two bugs in the netcdf-c cmake build that we identified and fixed in our fork.

1. `ncgen` is missing rpaths for hdf5 and won't run - see https://github.com/JCSDA/spack-packages/commit/27bc9e51bfd45064c965b9253f64f88640b34acd
2. Bug fixto find MPI functions when build system is cmake. See https://github.com/Unidata/netcdf-c/issues/3199 for details. The solution adopted here is the same as the suggestion made in https://github.com/Unidata/netcdf-c/issues/3199, which looks like will make its way into the authoritative codebase soon

Note. This patch is admittedly a bit of a hack. But given the overall state of the `netcdf-c` CMake build system I believe that this duct-tape solution is good enough. I hope that a overhaul of the entire CMake build for the package will solve this problem in a better way.

## Testing

Both patches were tested in https://github.com/JCSDA/spack-stack branch `feature/update_to_spack_v1` for `intel-oneapi-compilers + intel-oneapi-mpi` and `gcc + openmpi`.

climbfuji committed:
Bug fix in repos/spack_repo/builtin/packages/netcdf_c/package.py to find MPI functions when build system is cmake

climbfuji committed:
For netcdf-c, apply cmake patches only of build_system=cmake

climbfuji commented:
"@when" is not working as expected, will fix together with the style errors.

climbfuji commented:
Reverting to draft while I am still working on this

climbfuji committed:
Updated version of 'MPI_{Comm,Info}_C' bug fix for netcdf-c; tested with intel-oneapi-compilers+intel-oneapi-mpi and gcc+openmpi

climbfuji committed:
Merge branch 'develop' of https://github.com/spack/spack-packages into bugfix/netcdf_c_cmake_build_spack_pkgs

climbfuji committed:
Fix 86952b15b516d8402a63cd53e955f7b6b0d1805f

climbfuji commented:
@skosukhin @wardf This is ready for your review. I found that the original logic to check if the function exists in CMake was not working reliably. It worked for gcc + openmpi when I included the MPI link libraries as in my original attempt. It didn't work for Intel OneAPI. And I think it wouldn't work for many other MPI providers, because there is no way add an `#include mpi.h` into the `CheckFunctionExists.c` cmake file that is used for the test.

I will give this a test on one of the Cray machines that I have access to, but the caveat is that the machine doesn't use they Cray compiler wrappers in our spack config (it uses the vendor compilers like on non-Cray machines).

climbfuji commented:
> @skosukhin @WardF This is ready for your review. I found that the original logic to check if the function exists in CMake was not working reliably. It worked for gcc + openmpi when I included the MPI link libraries as in my original attempt. It didn't work for Intel OneAPI. And I think it wouldn't work for many other MPI providers, because there is no way add an `#include mpi.h` into the `CheckFunctionExists.c` cmake file that is used for the test.
> 
> I will give this a test on one of the Cray machines that I have access to, but the caveat is that the machine doesn't use they Cray compiler wrappers in our spack config (it uses the vendor compilers like on non-Cray machines).

Update. I tested this on a number of machines, including Cray systems with at least two compilers (GNU and Intel LLVM). For each system/compiler/mpi combination the symbols `MPI_{Comm,Info}_f2c` are found with this patch.

climbfuji commented:
Pinging the reviewers to take a look at this PR, please. Thanks!

climbfuji committed:
Merge branch 'develop' into bugfix/netcdf_c_cmake_build_spack_pkgs

climbfuji commented:
> Pinging the reviewers to take a look at this PR, please. Thanks!

Pinging all the reviewers once more before asking the spack developers to proceed w/o approval. Thanks very much in advance!

@skosukhin @WardF @johnwparent

climbfuji committed:
Merge branch 'develop' into bugfix/netcdf_c_cmake_build_spack_pkgs

johnwparent commented on repos/spack_repo/builtin/packages/netcdf_c/package.py:
```diff
@@ -143,6 +143,20 @@ class NetcdfC(CMakePackage, AutotoolsPackage):
+        if self.spec.satisfies("build_system=cmake"):
+            filter_file(
+                "SET(CMAKE_INSTALL_RPATH_USE_LINK_PATH FALSE)",
+                "#SET(CMAKE_INSTALL_RPATH_USE_LINK_PATH FALSE)",
```
If you want this enabled, I'd explicitly set this to true, i.e.

`set(CMAKE_INSTALL_RPATH_USE_LINK_PATH TRUE)`

Otherwise this could still be false unexpectedly. This doesn't stop other places in the code from changing the setting, but explicitly setting it here guarantees you'll be getting the expected settings for the targets you care about.

climbfuji replied:
Good point, happy to make this change.

johnwparent commented on repos/spack_repo/builtin/packages/netcdf_c/package.py:
```diff
@@ -379,6 +393,8 @@ def cmake_args(self):
             self.define_from_variant("NETCDF_ENABLE_LOGGING", "logging"),
         ]
+        if any(self.spec.satisfies(s) for s in ["+mpi", "+parallel-netcdf", "^hdf5+mpi~shared"]):
+            base_cmake_args.append(self.define("CMAKE_C_COMPILER", self.spec["mpi"].mpicc))
```
Should this be done for CXX as well? 

Why just `hdf~shared`?

climbfuji replied:
I think the condition in the line above
```
if any(self.spec.satisfies(s) for s in ["+mpi", "+parallel-netcdf", "^hdf5+mpi~shared"]):
```
evaluates to `True` if any (just one is enough) of these conditions is met.

climbfuji replied:
FWIW, I copied this line from the autotools builder to the cmake builder.

climbfuji replied:
See line 512

johnwparent replied:
Should have been clearer, sorry. 
I meant why hdf5 `~shared` and not shared or static? 
This is a check to add the MPI C compiler wrapper if we're building with MPI, so I'm wondering why we'd only want to do this for static hdf5?

climbfuji replied:
The answer is still the same - because that's how it was done for the autotools builder by someone else long before I touched the code the first time :-)

My guess is that for a shared build, the required libraries are dynamically linked to the hdf5 shared library, while for a static hdf5 the MPI libraries need to be added explicitly.

johnwparent replied:
Heh, fair enough, I'm still suspicious as CMake and autotools handle things like this quite differently, but I'd have to look deeper and this is fine for now.

johnwparent replied:
It does look like gitlab is failing on Windows, and given that it's a compiler detection issue, this change is potentially the source (I think you might need to also specify the CXX compiler).

If you don't have a Windows box to debug on let me know and I can figure out what's failing.

climbfuji replied:
I don't have a windows machine to debug this; therefore I'd appreciate your help. Thank you!

climbfuji committed:
Update package.py

Explicitly set CMAKE_INSTALL_RPATH_USE_LINK_PATH to TRUE

climbfuji commented on repos/spack_repo/builtin/packages/netcdf_c/package.py:
```diff
@@ -143,6 +143,20 @@ class NetcdfC(CMakePackage, AutotoolsPackage):
+        if self.spec.satisfies("build_system=cmake"):
+            filter_file(
+                "SET(CMAKE_INSTALL_RPATH_USE_LINK_PATH FALSE)",
+                "SET(CMAKE_INSTALL_RPATH_USE_LINK_PATH TRUE)",
```
@johnwparent done in 153371484b01cc2e12525202a37316009894f219

climbfuji committed:
Merge branch 'develop' into bugfix/netcdf_c_cmake_build_spack_pkgs

climbfuji committed:
Merge branch 'develop' into bugfix/netcdf_c_cmake_build_spack_pkgs

climbfuji commented:
@johnwparent Is there anything I can do to move this PR forward? Thanks!

climbfuji committed:
Merge branch 'develop' into bugfix/netcdf_c_cmake_build_spack_pkgs

johnwparent commented:
> @johnwparent Is there anything I can do to move this PR forward? Thanks!

Nothing from your end, more a bandwidth issue from my end, will be getting to this shortly!

climbfuji commented:
> > @johnwparent Is there anything I can do to move this PR forward? Thanks!
> 
> Nothing from your end, more a bandwidth issue from my end, will be getting to this shortly!

Thanks, very much appreciated!

climbfuji commented:
@johnwparent Did you have a chance to look at this? Thank you.

johnwparent committed:
NetCDF-c: Ensure plugin path is posix

Otherwise CMake will choke on a Windows escape character if based in the C: drive

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent commented:
> @johnwparent Did you have a chance to look at this? Thank you.

Thanks for the ping, resolved by #4293

climbfuji commented:
> > @johnwparent Did you have a chance to look at this? Thank you.
> 
> Thanks for the ping, resolved by #4293

This PR does more/other stuff than #4293. Which part of this PR #2384 is resolved by #4293?

johnwparent commented:
> > > @johnwparent Did you have a chance to look at this? Thank you.
> > 
> > 
> > Thanks for the ping, resolved by #4293
> 
> This PR does more/other stuff than #4293. Which part of this PR #2384 is resolved by #4293?

Sorry, should have been clear, the CI failure I promised to look into is fixed by that PR

johnwparent committed:
[@spackbot] updating style on behalf of johnwparent

climbfuji commented:
> > > > @johnwparent Did you have a chance to look at this? Thank you.
> > > 
> > > 
> > > Thanks for the ping, resolved by #4293
> > 
> > 
> > This PR does more/other stuff than #4293. Which part of this PR #2384 is resolved by #4293?
> 
> Sorry, should have been clear, the CI failure I promised to look into is fixed by that PR

Ahh, that makes sense. I'll wait for #4293 to be merged then. Thanks!

climbfuji committed:
Merge branch 'develop' into bugfix/netcdf_c_cmake_build_spack_pkgs

climbfuji committed:
Merge branch 'develop' of https://github.com/spack/spack-packages into bugfix/netcdf_c_cmake_build_spack_pkgs

climbfuji committed:
Merge branch 'fixup-netcdf-plugin-prefix-path' of https://github.com/johnwparent/spack-packages into bugfix/netcdf_c_cmake_build_spack_pkgs

climbfuji commented:
@johnwparent I pulled in your changes from https://github.com/spack/spack-packages/pull/4293 and the basic unit tests now pass for Windows. The gitlab-ci tests are still being queued.

climbfuji commented:
@johnwparent What do we need to do to move your PR forward so that I can update this PR and we can review & merge?

johnwparent commented:
> @johnwparent What do we need to do to move your PR forward so that I can update this PR and we can review & merge?

@climbfuji I just need to be able to successfully reproduce the paraview build failure locally. So far no dice. I'll let you know as soon as that's ready

johnwparent commented:
Blocker fix is merged, let's rebase and LGTM

climbfuji committed:
Merge branch 'develop' of https://github.com/spack/spack-packages into bugfix/netcdf_c_cmake_build_spack_pkgs

climbfuji commented:
> Blocker fix is merged, let's rebase and LGTM

Thanks, done! Please take another look.

climbfuji commented:
This error is new, I haven't changed anything, though, just pulled in develop:
```
spack.repo.RepoError: cannot load package 'netcdf-c' from the 'builtin' repository: NetcdfC.patch() got an unexpected keyword argument 'sha256'
```

climbfuji committed:
Merge branch 'develop' of https://github.com/spack/spack-packages into bugfix/netcdf_c_cmake_build_spack_pkgs

climbfuji committed:
Fix for patch sha256 error: move 'def patch' after all other patch() calls

climbfuji committed:
Merge branch 'develop' of https://github.com/spack/spack-packages into bugfix/netcdf_c_cmake_build_spack_pkgs

johnwparent reviewed (approved):
Looks good, thanks!

climbfuji commented:
@johnwparent We are so close, what do we need to do to merge this? Thanks!

johnwparent commented:
> @johnwparent We are so close, what do we need to do to merge this? Thanks!

Looks like gitlab CI failed and never reported. Seeing the Windows netcdf-c failure we fixed earlier... looking into it

climbfuji commented:
> > @johnwparent We are so close, what do we need to do to merge this? Thanks!
> 
> Looks like gitlab CI failed and never reported. Seeing the Windows netcdf-c failure we fixed earlier... looking into it

Thanks, and sorry that this update is causing so much work.

johnwparent commented:
> > > @johnwparent We are so close, what do we need to do to merge this? Thanks!
> > 
> > 
> > Looks like gitlab CI failed and never reported. Seeing the Windows netcdf-c failure we fixed earlier... looking into it
> 
> Thanks, and sorry that this update is causing so much work.

NW, it's more annoying these spurious windows failures are blocking you and no one else :/

I can't replicate the netcdf-c failure locally off develop (both core and packages). Can you try rebasing one more time?

climbfuji committed:
Merge branch 'develop' into bugfix/netcdf_c_cmake_build_spack_pkgs

climbfuji commented:
Sure, no problem. Fingers crossed!

climbfuji commented:
@johnwparent Looks like this didn't make it into the release :-(

climbfuji committed:
Merge branch 'develop' into bugfix/netcdf_c_cmake_build_spack_pkgs

bernhardkaindl committed:
Merge branch 'develop' into bugfix/netcdf_c_cmake_build_spack_pkgs

climbfuji commented on repos/spack_repo/builtin/packages/netcdf_c/cmakelists_mpi_symbols.patch:
```diff
@@ -0,0 +1,16 @@
+ CHECK_FUNCTION_EXISTS(gettimeofday  HAVE_GETTIMEOFDAY)
+-CHECK_FUNCTION_EXISTS(MPI_Comm_f2c  HAVE_MPI_COMM_F2C)
+-CHECK_FUNCTION_EXISTS(MPI_Info_f2c  HAVE_MPI_INFO_F2C)
++if(MPI_C_COMPILER_INCLUDE_DIRS)
++  set (CMAKE_REQUIRED_INCLUDES "${MPI_C_COMPILER_INCLUDE_DIRS}")
++endif()
++CHECK_SYMBOL_EXISTS(MPI_Comm_f2c "mpi.h" HAVE_MPI_COMM_F2C)
++CHECK_SYMBOL_EXISTS(MPI_Info_f2c "mpi.h" HAVE_MPI_INFO_F2C)
```
@bernhardkaindl @johnwparent I am ok with this change, but I don't think it's needed. We've used my original change on many systems with different MPI distributions (intel-oneapi-mpi, mpich, openmpi, cray-mpich).

bernhardkaindl replied:
@climbfuji, thanks for the helpful comment!

This instills a lot of confidence in the PR.

We'll have to have passiing GitLab CI before it can be merged.

I'll take a look at the failed GitLab pipeline and maybe try a rebase or a merge of develop to re-run the pipelines with the current develop.

But otherwise, unless there are further commentsI'd plan to approve and merge soon.

Thanks,
Bernhard

johnwparent replied:
This variable is set by the find mpi module (which is why its not used elsewhere), but it does mean its the location that the find module itself set the headers to be, so 
```
 so this logic is likely to be a no-op on many CMake/FindMPI setups and can leave CHECK_SYMBOL_EXISTS(... "mpi.h" ...) unable to find mpi.h
``` 
is categorically untrue (also obviously it works, checkmate copilot)

While I don't normally love using cache variables like this directly, and there are more user facing variables that would be better to use, this isn't technically _wrong_. 

We should use one of the intentionally user facing variables, but since that requires re-writing the patch a bit, I wasn't going to ask for it.

climbfuji replied:
Thanks.

This PR is becoming quite a thing. Process getting in the way of progress I suppose.

johnwparent replied:
It's just a bunch of flakey builds, straightforward in principle, time consuming and painstaking in practice.

bernhardkaindl commented:
The GitLab CI pipeline failed due to missing metadata, so that's unlikely due to changes in this PR.

Hoping to resovle the error, I merged develop to re-run the pipelines with the current develop (rebasing wasn't possible automatically on the web due to conflicts).

If the rebase passes CI, unless there are further comments, I'd assume we can approve and merge.

bernhardkaindl commented:
GitLab CI was failing due to an update if ispc@1.30 #4441 which does not build in CI
because GitLab CI does not have 32-bit development support and ispc@1.30 assumes it is present.
https://gitlab.spack.io/spack/spack-packages/-/jobs/23299494

As a PR for ispc itself apparently does not build it (or CI changes since the last build of #4441 in CI), this PR needs to test-drive (and have) the fix as CI won't pass without a fix for it. Pushed a tentative fix.

bernhardkaindl commented:
Due to spack not finding the the C++ of MSVC, I want to ensure a good base and rebased onto the current develop branch with manual conflict resolution. The new base commit of the PR is:
> py-pspamm: Use spack's defaullt to install the package (#5363) - 52ebb1821e80c271799bc590700dc151c370e321
```py
$ . ${SPACK_CI_SPACK_ROOT}\share\spack\setup-env.ps1
Invalid macro definition.
*****************************************************************
**************** Spack Package Manager **************************
*****************************************************************
$ If (Test-Path -path C:\\key\intermediate_ci_signing_key.gpg) { spack.ps1 gpg trust -y C:\\key\intermediate_ci_signing_key.gpg }
$ If (Test-Path -path C:\\key\spack_public_key.gpg) { spack.ps1 gpg trust -y C:\\key\spack_public_key.gpg }
$ spack.ps1 env activate --without-view ${SPACK_CONCRETE_ENV_DIR}
$ spack.ps1 config add "config:install_tree:projections:${SPACK_JOB_SPEC_PKG_NAME}/${SPACK_JOB_SPEC_DAG_HASH}:'morepadding/{hash}'"
$ mkdir ${SPACK_ARTIFACTS_ROOT}/user_data
    Directory: C:\builds\spack\spack-packages\jobs_scratch_dir\windows-vis
Mode                 LastWriteTime         Length Name                         
----                 -------------         ------ ----                         
d-----         7/11/2026  12:32 AM                user_data                    
$ spack.ps1 --backtrace ci rebuild | Tee-Object -FilePath "${env:SPACK_ARTIFACTS_ROOT}/user_data/pipeline_out.txt" 2>&1 | Tee-Object -FilePath "${env:SPACK_ARTIFACTS_ROOT}/user_data/pipeline_err.txt"
==> Warning: "git@2.45.2.windows.1~tcltk" detected in "C:\Program Files\Git\cmd" was already detected in "C:\Program Files\Git"
==> Warning: error detecting "msvc" from prefix C:\Program Files (x86)\Intel\oneAPI\compiler\2024.1\bin: Traceback (most recent call last):
  File "C:\builds\spack\spack-packages\.ci\tmp\spack\lib\spack\spack\detection\path.py", line 283, in detect_specs
    pkg.determine_spec_details(candidate_path, items_in_prefix)
  File "C:\builds\spack\spack-packages\.ci\tmp\spack\lib\spack\spack\package_base.py", line 218, in determine_spec_details
    variants = cls.determine_variants(objs, version_str)
  File "C:\builds\spack\spack-packages\repos\spack_repo\builtin\packages\msvc\package.py", line 77, in determine_variants
    extras["compilers"]["c"] = extras["compilers"]["cxx"]
KeyError: 'cxx'
System.Management.Automation.RemoteException
==> Compilers have been configured automatically from PATH inspection
```

Hm, the clean rebase on top of develop (gitlab used 8610e4bec971828a60d1ea1520c4572d0c6b020f) didn't help either. I guess we might need to ask the GitLab CI maintainers, I think.

climbfuji commented:
Thanks for your efforts @bernhardkaindl

johnwparent commented:
@bernhardkaindl the person you're looking to ask is the Windows maintainer, which is me :) 

You're chasing an unrelated non error right now. Spack finds the compilers, it just fails to find a CXX compiler for that particular detected compiler (which is correct, it's a fotran compiler with no cxx). We have a valid compiler layout in CI for Windows, this is just a quirk of how we modeled intel fortran + msvc before compilers as nodes and compiler mixing were allowed.

The actual error, if you scroll to the build output, is a "bad unicode escape" due to Windows `\` path separators. CMake does not like unescaped `\` and it seems as though one has snuck in here.

I'm working on this locally but I updated VS so I have a ton of failures related to that I need to work through.

climbfuji commented:
> @bernhardkaindl the person you're looking to ask is the Windows maintainer, which is me :)
> 
> You're chasing an unrelated non error right now. Spack finds the compilers, it just fails to find a CXX compiler for that particular detected compiler (which is correct, it's a fotran compiler with no cxx). We have a valid compiler layout in CI for Windows, this is just a quirk of how we modeled intel fortran + msvc before compilers as nodes and compiler mixing were allowed.
> 
> The actual error, if you scroll to the build output, is a "bad unicode escape" due to Windows `\` path separators. CMake does not like unescaped `\` and it seems as though one has snuck in here.

Please let me know if this is something I can/need to fix. Thanks!

johnwparent commented on repos/spack_repo/builtin/packages/netcdf_c/package.py:
```diff
@@ -395,6 +416,8 @@ def cmake_args(self):
             self.define_from_variant("NETCDF_ENABLE_LOGGING", "logging"),
         ]
+        if any(self.spec.satisfies(s) for s in ["+mpi", "+parallel-netcdf", "^hdf5+mpi~shared"]):
+            base_cmake_args.append(self.define("CMAKE_C_COMPILER", self.spec["mpi"].mpicc))
```
Ahh, I think this is our error on Windows. I can't confirm locally due to... a lot of unrelated build errors, but we're passing a path directly to CMake that is a Windows path. Let's wrap this in a pathlib Path and call `to_posix` and then cast to string

climbfuji replied:
Can you please suggest the changes you would like to make?

johnwparent replied:
I just pushed it myself

climbfuji replied:
Thanks very much, fingers crossed it all works now.

johnwparent replied:
netcdf-c passed! The rest should build

johnwparent commented:
@spackbot fix style

climbfuji commented:
All tests passed! Can one of you merge @johnwparent @bernhardkaindl ?

climbfuji commented:
Thanks very much @johnwparent for all your efforts on this.
